<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Работа-со-столбцами,-имеющими-пропуски" data-toc-modified-id="Работа-со-столбцами,-имеющими-пропуски-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Работа со столбцами, имеющими пропуски</a></span><ul class="toc-item"><li><span><a href="#Устранение-пропусков-в-столбце-'total_income'." data-toc-modified-id="Устранение-пропусков-в-столбце-'total_income'.-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Устранение пропусков в столбце 'total_income'.</a></span></li><li><span><a href="#Работа-с-аномалиями-и-устранение-пропусков-в-столбце-'days_employed'." data-toc-modified-id="Работа-с-аномалиями-и-устранение-пропусков-в-столбце-'days_employed'.-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>Работа с аномалиями и устранение пропусков в столбце 'days_employed'.</a></span><ul class="toc-item"><li><span><a href="#Рассмотрение-категорий-клиентов-по-типу-занятости,--для-которых-стаж-указан-как-положительное-число:" data-toc-modified-id="Рассмотрение-категорий-клиентов-по-типу-занятости,--для-которых-стаж-указан-как-положительное-число:-2.1.2.1"><span class="toc-item-num">2.1.2.1&nbsp;&nbsp;</span>Рассмотрение категорий клиентов по типу занятости,  для которых стаж указан как положительное число:</a></span></li><li><span><a href="#Определение-причин-наличия-отрицательных-значений-в-столбце-'days_employed'." data-toc-modified-id="Определение-причин-наличия-отрицательных-значений-в-столбце-'days_employed'.-2.1.2.2"><span class="toc-item-num">2.1.2.2&nbsp;&nbsp;</span>Определение причин наличия отрицательных значений в столбце 'days_employed'.</a></span></li><li><span><a href="#Определение-алгоритма-работы-с-пропусками-в-столбце-'days_employed'." data-toc-modified-id="Определение-алгоритма-работы-с-пропусками-в-столбце-'days_employed'.-2.1.2.3"><span class="toc-item-num">2.1.2.3&nbsp;&nbsp;</span>Определение алгоритма работы с пропусками в столбце 'days_employed'.</a></span></li></ul></li></ul></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span><ul class="toc-item"><li><span><a href="#Проверка-наличия-аномалий-в-столбце-'children'." data-toc-modified-id="Проверка-наличия-аномалий-в-столбце-'children'.-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Проверка наличия аномалий в столбце 'children'.</a></span></li><li><span><a href="#Проверка-наличия-аномалий-в-столбце-'family_status'-('family_status_id')." data-toc-modified-id="Проверка-наличия-аномалий-в-столбце-'family_status'-('family_status_id').-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Проверка наличия аномалий в столбце 'family_status' ('family_status_id').</a></span></li><li><span><a href="#Проверка-наличия-аномалий-в-столбце-'total_income'." data-toc-modified-id="Проверка-наличия-аномалий-в-столбце-'total_income'.-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>Проверка наличия аномалий в столбце 'total_income'.</a></span></li></ul></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Удаление дубликатов.</a></span><ul class="toc-item"><li><span><a href="#Приведение-к-единому-регистру-строковых-значений" data-toc-modified-id="Приведение-к-единому-регистру-строковых-значений-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>Приведение к единому регистру строковых значений</a></span></li><li><span><a href="#Удаление-явных-дубликатов" data-toc-modified-id="Удаление-явных-дубликатов-2.4.2"><span class="toc-item-num">2.4.2&nbsp;&nbsp;</span>Удаление явных дубликатов</a></span></li></ul></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:" data-toc-modified-id="Вопрос-1:-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вопрос 1:</a></span><ul class="toc-item"><li><span><a href="#Вывод-1:" data-toc-modified-id="Вывод-1:-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Вывод 1:</a></span></li></ul></li><li><span><a href="#Вопрос-2:" data-toc-modified-id="Вопрос-2:-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Вопрос 2:</a></span><ul class="toc-item"><li><span><a href="#Вывод-2:" data-toc-modified-id="Вывод-2:-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Вывод 2:</a></span></li></ul></li><li><span><a href="#Вопрос-3:" data-toc-modified-id="Вопрос-3:-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Вопрос 3:</a></span><ul class="toc-item"><li><span><a href="#Вывод-3:" data-toc-modified-id="Вывод-3:-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Вывод 3:</a></span></li></ul></li><li><span><a href="#Вопрос-4:" data-toc-modified-id="Вопрос-4:-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вопрос 4:</a></span><ul class="toc-item"><li><span><a href="#Вывод-4:" data-toc-modified-id="Вывод-4:-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Вывод 4:</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков.

Необходимо разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.
По результатам исследования, требуется ответить на следующие вопросы:
* Есть ли зависимость между количеством детей и возвратом кредита в срок?
* Есть ли зависимость между семейным положением и возвратом кредита в срок?
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
* Как разные цели кредита влияют на его возврат в срок?

## Обзор данных

Импортируем библиотеки Pandas, Numpy:

In [1]:
import pandas as pd
import numpy as np

Откроем представленные входные данные от банка с рассмотрением общей информации о данных и выводом первых 10 строк таблицы (представленным данным присвоим имя df):

In [2]:
df = pd.read_csv('C:/projects/project_2/data.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Исходя из полученной документации имеется следующая информация о наименованиях столбцов:
- **children** — количество детей в семье;
- **days_employed** — общий трудовой стаж в днях;
- **dob_years** — возраст клиента в годах;
- **education** — уровень образования клиента;
- **education_id** — идентификатор уровня образования;
- **family_status** — семейное положение;
- **family_status_id** — идентификатор семейного положения;
- **gender** — пол клиента;
- **income_type** — тип занятости;
- **debt** — имел ли задолженность по возврату кредитов;
- **total_income** — ежемесячный доход;
- **purpose** — цель получения кредита.

По результатам рассмотрения полученной информации о данных, отмечаем следующее:
- имеются пропуски в столбце 'days_employed';
- имеются пропуски в столбце 'total_income';
- данные столбца 'education' заполнены различным регистром, но при этом имеется столбец education_id (предварительно можно сказать (в дальнейшем требует проверки), что '0' - при наличии высшего, '1' - при наличии среднего образования);
- данные в столбце 'days_employed' имеют частично отрицательные значения, что не соответствует описанию столбца

## Предобработка данных

### Работа со столбцами, имеющими пропуски

В имеющихся данных пропуски присутствуют в двух столбцах: 'days_employed', 'total_income'. Так как целями исследования определена необходимость определения зависимости возврата от уровня дохода, целесообразно начать устранение пропусков со столбца 'total_income'.

#### Устранение пропусков в столбце 'total_income'.

Предварительно рассмотрим выборку первых пяти строк с пропущенными значениями в столбце 'total_income':

In [4]:
df[df['total_income'].isna()].head()

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

Строки с пропусками в столбце  'total_income' в выведенной части таблицы имеют значение NaN также и в столбце 'days_employed'. Проверим, соответствуют ли пропуски в ячейках указанных столбцов для остального датафрейма: 

In [5]:
print('\n',
    'Количество пропусков в столбце \'total_income\' =', df['total_income'].isna().sum(),'\n',
    'Доля пропусков в столбце \'total_income\' от общего количества строк =', df['total_income'].isna().mean(),'\n', 
    'Количество пропусков в столбце \'days_employed\' =', df['days_employed'].isna().sum(),'\n',
    'Количество строк с пропусками по всей таблице =', len(df[df.isna().any(1)]))


 Количество пропусков в столбце 'total_income' = 2174 
 Доля пропусков в столбце 'total_income' от общего количества строк = 0.10099883855981417 
 Количество пропусков в столбце 'days_employed' = 2174 
 Количество строк с пропусками по всей таблице = 2174


Количество строк, в которых есть пропущенные значения для столбцов 'total_income', 'days_employed' и всей таблицы одинаково, что говорит о том, что указанные столбцы имеют пропуски в одних и тех же строках. 

С целью определения целесообразного способа заполнения пропущенных значений, рассмотрим в каком диапазоне лежат значения ячеек, а также среднее и медианное значения столбца:

In [6]:
print('\n',
    'Минимальное значение в столбце \'total_income\' =', min(df['total_income']),'\n',
    'Максимальное значение в столбце \'total_income\' =', max(df['total_income']),'\n',
    'Среднее значение в столбце \'total_income\' =', df['total_income'].mean(),'\n',
    'Медианное значение в столбце \'total_income\' =', df['total_income'].median(),'\n',)


 Минимальное значение в столбце 'total_income' = 20667.26379327158 
 Максимальное значение в столбце 'total_income' = 2265604.028722744 
 Среднее значение в столбце 'total_income' = 167422.30220817294 
 Медианное значение в столбце 'total_income' = 145017.93753253992 



Так как значения ячеек столбца 'total_income' различаются более чем на 2 порядка, целесообразно заполнять пропущенные значения медианными:

In [7]:
df['total_income'] = df['total_income'].fillna(df['total_income'].median())

Убедимся в отсутствии пропусков в столбце:

In [8]:
df['total_income'].isna().sum()

0

#### Работа с аномалиями и устранение пропусков в столбце 'days_employed'.

В соответствии с целью выполнения работ в настоящем проекте, данные столбца 'days_employed' находятся вне границ заданных вопросов. Однако указанный столбец содержит следующие аномалии, причины которых целесообразно определить (и устранить) для возможности работы с данными в дальнейшем при необходимости:
- положительные значения слишком велики (в первых 10 строках таблицы имеется одно положительное значение, равное 340266, т.е. почти 1000 лет);
- ряд значений отрицательны (очевидно, что трудовой стаж не может быть отрицательным);
- в столбце ряд строк имеет пропуски (пропуски имеются в тех же строках, в которых пропущены значения в столбце 'total_income').

##### Рассмотрение категорий клиентов по типу занятости,  для которых стаж указан как положительное число:

In [9]:
days_pos = df.loc[df['days_employed']>0]
days_pos.groupby('income_type')['days_employed'].agg(['count','min','mean','max'])

count            min           mean            max
income_type                                                    
безработный      2  337524.466835  366413.652744  395302.838654
пенсионер     3443  328728.720605  365003.491245  401755.400475

Т.е. строки, в которых в качестве стажа указано положительное число (все положительные числа имеют величину, значительно превышающую продолжительность жизни), относятся к клиентов двух типов (по типу занятости): безработные и пенсионеры. <br>Рассмотрим, сколько всего строк относятся к занятости данного типа и количество пропусков в них (при необходимости):

In [10]:
df.loc[df['income_type']=='пенсионер']['income_type'].count()

3856

In [11]:
df.loc[df['income_type']=='безработный']['income_type'].count()

2

In [12]:
df.loc[df['income_type']=='пенсионер']['days_employed'].isna().agg(['sum','mean'])

sum     413.000000
mean      0.107106
Name: days_employed, dtype: float64

**Исходя из полученной информации, очевиден следующий вывод: большие положительные значения стажа являются результатом автозаполнения для клиентов, не имеющих текущий трудовой стаж. Выполним замену данных значений на 0:** 

In [13]:
df.loc[(df['days_employed']>0), 'days_employed'] = 0

##### Определение причин наличия отрицательных значений в столбце 'days_employed'.

Рассмотрим сгруппированную по типу занятости таблицу, в которых стаж указан как отрицательное число:

In [14]:
days_pos = df.loc[df['days_employed']<0]
days_pos.groupby('income_type')['days_employed'].agg(['count','min','mean','max'])

count           min         mean          max
income_type                                                   
в декрете            1  -3296.759962 -3296.759962 -3296.759962
госслужащий       1312 -15193.032201 -3399.896902   -39.954170
компаньон         4577 -17615.563266 -2111.524398   -30.195337
предприниматель      1   -520.848083  -520.848083  -520.848083
сотрудник        10014 -18388.949901 -2326.499216   -24.141633
студент              1   -578.751554  -578.751554  -578.751554

**Очевидно, что отрицательное значение стажа в днях является результатом автоматического определения стажа исходя из даты начала трудовой деятельности (т.е. клиентем заполняется дата начала трудовой деятельности, а значение в столбец записывается исходя из даты его формирования и т.к. дата является прошедшим событием, то число имеет знак '-'). Заменим отрицательные значения на их значения по модулю:**

In [15]:
df['days_employed'] = df['days_employed'].abs()

##### Определение алгоритма работы с пропусками в столбце 'days_employed'.

**Заполнение пропущенных значений столбца 'days_employed' целесообразно выполнять следующим образом:**
- **часть пропусков (порядка 11%) в строках клиентов с категорией занятости 'пенсионер', целесообразно заполнять нулевыми значениями;**
- **остальные пропуски целесообразно заполнять медианным значением по столбцу 'days_employed'.** 
<br>Выполним указанное заполнение:

In [16]:
df.loc[df['income_type']=='пенсионер', 'days_employed'] = 0

In [17]:
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())

Для удобства дальнейше работы (если будет такая необходимость), приведем тип данных столбца к целочисленному:

In [18]:
df = df.astype({'days_employed': 'int64'})

### Проверка данных на аномалии и исправления.

Исходя из цели проведения работы, необходима проверка наличия аномалий в следующих столбцах:
- 'children' — количество детей в семье;
- 'family_status' — семейное положение (в том числе совместно со столбцом 'family_status_id');
- 'total_income' — ежемесячный доход;
- 'purpose' — цель получения кредита.

#### Проверка наличия аномалий в столбце 'children'.

Выведем перечень значений столбца 'chilren', встречающийся в таблице:

In [19]:
df['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

Распределение клиентов по количеству детей:

In [20]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

**Представленное распределение показывает, что имеются очевидно ошибочные данные в столбце: -1 (значение не может быть отрицательным), 20 (76 случаев семей с количеством в 20 детей при отстутсвии строк мс количеством между 5 и 20 также очевидно неверно). Заменим ошибочно внесенные данные на медианные:**

In [21]:
df['children'].median()

0.0

In [22]:
df.loc[(df['children']==20), 'children'] = df['children'].median()
df.loc[(df['children']==-1), 'children'] = df['children'].median()

Проверим результативность выполненной операции:

In [23]:
df['children'].unique()

array([1., 0., 3., 2., 4., 5.])

Вернем значения столбца к целочисленному виду: 

In [24]:
df = df.astype({'children': 'int64'})

#### Проверка наличия аномалий в столбце 'family_status' ('family_status_id').

Выведем перечень значений столбца 'family_status' ('family_status_id'), встречающийся в таблице:

In [25]:
print(df['family_status'].unique(), df['family_status_id'].unique())

['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'Не женат / не замужем'] [0 1 2 3 4]


In [26]:
df['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

Рассмотрим, как эти значения соотносятся между собой (с оценкой их распределения):

In [27]:
family_table = pd.pivot_table(df,values='income_type', index=['family_status','family_status_id'],aggfunc='count')

In [28]:
family_table

income_type
family_status         family_status_id             
Не женат / не замужем 4                        2813
в разводе             3                        1195
вдовец / вдова        2                         960
гражданский брак      1                        4177
женат / замужем       0                       12380

**Аномалии в столбце 'family_status' ('family_status_id') не выявлены.**

#### Проверка наличия аномалий в столбце 'total_income'.

Диапазон значений столбца 'total_income' рассматривался в п. 2.1.1, в соответстии с которым аномалии выявлены не были. <br>Рассмотрим распределение среднего дохода в зависимости от типа занятости:

In [29]:
df.groupby('income_type')['total_income'].agg(['count','min','mean','max'])

count            min           mean           max
income_type                                                       
безработный          2   59956.991984  131339.751676  2.027225e+05
в декрете            1   53829.130729   53829.130729  5.382913e+04
госслужащий       1459   29200.077193  168290.760408  9.104515e+05
компаньон         5085   28702.812889  196683.153073  2.265604e+06
пенсионер         3856   20667.263793  137972.581061  7.351033e+05
предприниматель      2  145017.937533  322090.541240  4.991631e+05
сотрудник        11119   21367.648356  159754.181986  1.726276e+06
студент              1   98201.625314   98201.625314  9.820163e+04

Аномалии в представленных значения не определены, однако целесообразно привести тип данных к целочисленным int для удобства дальнейшей работы:

In [30]:
df = df.astype({'total_income': 'int64'})

Проверим корректность выполненого действия:

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


In [32]:
df.groupby('income_type')['total_income'].agg(['count','min','mean','max'])

count     min           mean      max
income_type                                           
безработный          2   59956  131339.000000   202722
в декрете            1   53829   53829.000000    53829
госслужащий       1459   29200  168290.205620   910451
компаньон         5085   28702  196682.615536  2265604
пенсионер         3856   20667  137972.038900   735103
предприниматель      2  145017  322090.000000   499163
сотрудник        11119   21367  159753.634589  1726276
студент              1   98201   98201.000000    98201

### Изменение типов данных.

Тип данных в столбце 'total_income' изменен в п. 2.2.3. Для других данных изменение типа не требуется

### Удаление дубликатов.

#### Приведение к единому регистру строковых значений

До проведения работы с дубликатами целесообразно приведение к единому регистру столбцы со строковыми данными. Из не рассмотренных ранее столбцов такими являются 'education' (вывод первых 10 строк уже показал наличие различий в регистре ввода данных) и 'purpose' (вывод первых 10 строк различий регистра не показали, однако требуется проверка).
<br>Рассмотрим перечень уникальных значений в столбце 'education':

In [33]:
df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

Приведем данные значения к единому регистру:

In [34]:
df['education'] = df['education'].str.lower()

Проверим, что значения ячеек теперь не имеют символов в разных регистрах:

In [35]:
df['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Рассмотрим перечень уникальных значений в столбце 'purpose':

In [36]:
df['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

**Изменений регистра не требуется.**

#### Удаление явных дубликатов

Посчитаем, сколько явных дубликатов присутствует в таблице:

In [37]:
df.duplicated().sum()

71

Удалим данные дубликаты со сбросом индекса:

In [38]:
df = df.drop_duplicates().reset_index(drop=True)

Проверим результативность выполненного действия:

In [39]:
df.duplicated().sum()

0

Среди возможных причин наличия указанных дубликатов - повторные представления, неправильное соединение данных из разных источников, ошибки пользователя при занесении информации.

### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

В таблице с данными представлены столбцы, информация которых дублирована -   'family_status' и 'family_status_id',  'education' и 'education_id'. Для удобства дальнейшей работы целесообразно создание датафреймов-словарей, которые будут отражать, какое значений в столбце "_id" соответствует 'family_status' и 'education' с последующим удалением этих столбцов из исходного датафрейма. <br>Сформируем датафрейм для 'family_status':

In [40]:
family_df = df[['family_status_id', 'family_status']].drop_duplicates().reset_index(drop=True)
family_df

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

 Создадим аналогичный датафрейм для 'education':

In [41]:
education_df = df[['education_id', 'education']].drop_duplicates().reset_index(drop=True)
education_df

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

Исключим столбцы 'family_status' и  'education' из исходной таблицы:

In [42]:
df = df.drop(columns=['family_status', 'education'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21454 non-null  int64 
 1   days_employed     21454 non-null  int64 
 2   dob_years         21454 non-null  int64 
 3   education_id      21454 non-null  int64 
 4   family_status_id  21454 non-null  int64 
 5   gender            21454 non-null  object
 6   income_type       21454 non-null  object
 7   debt              21454 non-null  int64 
 8   total_income      21454 non-null  int64 
 9   purpose           21454 non-null  object
dtypes: int64(7), object(3)
memory usage: 1.6+ MB


### Категоризация дохода.

Для дальнейшей работы целесообразно следующее разделение клинетов на категории в зависимости от уровня дохода:
- клиенты с доходом до 30000 - категория 'Е';
- клиенты с доходом от 30001 до 50000 - категория 'D'
- клиенты с доходом от 50001 до 200000 - категория 'C'
- клиенты с доходом от 200001 до 1000000 - категория 'B'
- клиенты с доходом от 1000001 - категория 'A.
<br>Выполним указанное распределение с последующим добавлением соответствующего столбца в исходную таблицу:

In [43]:
def cat_income(total_income):
    if total_income <=30000:
        return 'E'
    if total_income <=50000:
        return 'D'
    if total_income <=200000:
        return 'C'
    if total_income <=1000000:
        return 'B'
    return 'A'
df['total_income_category'] = df['total_income'].apply(cat_income)
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0              0         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

### Категоризация целей кредита.

С целью категоризации значений в столбце 'purpose', целесообразно лемматизировать значения строк в столбце для дальнейшего разделения на категории. Выведеме перечень уникальных значений столбца 'purpose':

In [44]:
purposes = df['purpose'].unique()
print(purposes)
len(purposes)

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


38

Создадим функцию, возвращающую категорию на основании данных столбца 'purpose' следующим образом:
- при наличии в строке 'автомоб' - возвращение 'операции с автомобилем';
- при наличии в строке 'недвижим' или 'жиль' - возвращение 'операции с недвижимостью';
- при наличии в строке 'свадьб' - возвращение 'проведение свадьбы';
- при наличии в строке 'образован' - возвращение 'получение образования';

Проверим работу цикла на списке уникальных элементов purposes (с контролем количества выведенных значений на предмет соответствия len(purposes)):

In [45]:
count = 0
for pur in purposes:
        if 'жиль' in pur:
            print('операции с недвижимостью')
            count+=1
        if 'недвижим' in pur:
            print('операции с недвижимостью')
            count+=1
        if 'автомоб' in pur:
            print('операции с автомобилем')
            count+=1
        if 'свадьб' in pur:
            print('проведение свадьбы')
            count+=1
        if 'образован' in pur:
            print('получение образования')
            count+=1
print(count)

операции с недвижимостью
операции с автомобилем
получение образования
проведение свадьбы
операции с недвижимостью
получение образования
проведение свадьбы
операции с недвижимостью
операции с недвижимостью
операции с недвижимостью
операции с недвижимостью
операции с недвижимостью
операции с недвижимостью
операции с недвижимостью
операции с автомобилем
операции с автомобилем
операции с недвижимостью
операции с недвижимостью
операции с недвижимостью
операции с недвижимостью
операции с автомобилем
получение образования
операции с автомобилем
получение образования
операции с автомобилем
проведение свадьбы
получение образования
операции с недвижимостью
операции с недвижимостью
получение образования
операции с автомобилем
операции с автомобилем
получение образования
получение образования
операции с недвижимостью
операции с автомобилем
операции с недвижимостью
получение образования
38


Создадим функцию pur_cat для категоризации столбца 'purpose':

In [46]:
def pur_cat (purpose):
        if 'жиль' in purpose:
            return 'операции с недвижимостью'
        if 'недвижим' in purpose:
            return 'операции с недвижимостью'
        if 'автомоб' in purpose:
            return 'операции с автомобилем'
        if 'свадьб' in purpose:
            return 'проведение свадьбы'
        if 'образован' in purpose:
            return 'получение образования' 
df['purpose_category'] = df['purpose'].apply(pur_cat)
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0              0         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

## Ответы на вопросы.

### Вопрос 1:

**Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Для ответа на данный вопрос построим таблицу, в которой будет указана доля имеющих долг клиентов с разбивкой по количеству детей:

In [47]:
df.groupby('children')['debt'].agg(['count','mean'])

count      mean
children                 
0         14214  0.075419
1          4808  0.092346
2          2052  0.094542
3           330  0.081818
4            41  0.097561
5             9  0.000000

#### Вывод 1:

**Представленная таблица показывает следующее: наибольшая доля должников среди клиентов, имеющих 4 детей (9,76%), наименьшая - среди клиентов без детей (7,54%) если не учитывать клиентов с 5 детей, количество которых статистически не значимок (9 из 21525). Таким образом, количество детей влияет на вероятность возврата кредита в срок.**

### Вопрос 2:

**Есть ли зависимость между семейным положением и возвратом кредита в срок?**

Для ответа на данный вопрос построим таблицу, в которой будет указана доля имеющих долг клиентов с разбивкой по типам семейного положения:

In [48]:
family = df.groupby('family_status_id')['debt'].agg(['count','mean'])
family

count      mean
family_status_id                 
0                 12339  0.075452
1                  4151  0.093471
2                   959  0.065693
3                  1195  0.071130
4                  2810  0.097509

Для наглядности добавим в таблицу значения ключей для категорий семейного положения:

In [49]:
family_full = family.merge(family_df, on='family_status_id', how='left')
family_full

family_status_id  count      mean          family_status
0                 0  12339  0.075452        женат / замужем
1                 1   4151  0.093471       гражданский брак
2                 2    959  0.065693         вдовец / вдова
3                 3   1195  0.071130              в разводе
4                 4   2810  0.097509  Не женат / не замужем

#### Вывод 2:

**Зависимость между семейным положением и возвратом кредита в срок следующая: наибольшая вероятность несвоевременного возврата (или невозврата) кредита в срок у клиентов, которые не женаты/не замужем (9,75%) или находящихся в гражданском браке (9,35%). Клиенты, находящиеся или находившиеся в браке имеют вероятность своевременного возврата кредита выше: доля вдовцов/вдов, не возвращающих кредити - 6,57%, разведенных - 7,11%, находящихся в браке - 7,54%.**

### Вопрос 3:

**Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

Для ответа на данный вопрос построим таблицу, в которой будет указана доля имеющих долг клиентов с разбивкой по категориям уровня дохода:

In [50]:
df.groupby('total_income_category')['debt'].agg(['count','mean'])

count      mean
total_income_category                 
A                         25  0.080000
B                       5041  0.070621
C                      16016  0.084915
D                        350  0.060000
E                         22  0.090909

- клиенты с доходом до 30000 - категория 'Е';
- клиенты с доходом от 30001 до 50000 - категория 'D'
- клиенты с доходом от 50001 до 200000 - категория 'C'
- клиенты с доходом от 200001 до 1000000 - категория 'B'
- клиенты с доходом от 1000001 - категория 'A.

#### Вывод 3:

**Зависимость между уровнем дохода и возвратом кредита в срок не очевидна: минимальная доля должников в категориях 'B' (7,06%) и 'D' (6%). Максимальная  - среди должников с наименьшим доходом (9,1%) и середняков (8,49%).**

### Вопрос 4:

**Как разные цели влияют на его возврат в срок?**

Для ответа на данный вопрос построим таблицу, в которой будет указана доля имеющих долг клиентов с разбивкой по категориям цели кредита:

In [51]:
purpose_table = pd.pivot_table(df,values='debt', index=['purpose_category'],aggfunc=['count','mean'])
purpose_table

count      mean
                           debt      debt
purpose_category                         
операции с автомобилем     4306  0.093590
операции с недвижимостью  10811  0.072334
получение образования      4013  0.092200
проведение свадьбы         2324  0.080034

#### Вывод 4:

**Наибольшую вероятность несвоевременного возврата (или невозврата) кредита имеют клиенты, которым кредит нужен на операции с автомобилем (9,36%) или получение образования (9,22%). Наименьший процент должников среди клиентов, которые берут кредит для операций с недвижимостью (7,23%).**

## Общий вывод:

Проведенный анализ данных по возврату кредита в срок среди клиентов банка показал следующее: наиболее надежными заемщиками являются следующие клиенты:
- не имеющие детей (доля должников - 7,54%);
- вдовцы/вдовы (6,57%);
- с доходом от 30001 до 50000 (6%);
- которым кредит необходим для проведения операций с недвижимостью (7,23%).
Наименее надежными заемщиками являются следующие клиенты:
- имеющие 4 детей (доля должников - 9,76%);
- неженатые или незамужние (9,75%);
- с доходом до 30000 (9,1%);
- которым кредит необходим для проведения операций с автомобилем (9,36%).**